### PR TITLE
add schema versioning for moray buckets

### PR DESCRIPTION
**Pull request by misterdjules**
This is part of the effort to split up the work that was done in https://github.com/joyent/sdc-vmapi/pull/16 into smaller changes.

This PR specifically adds versioning to VMAPI's moray buckets. It allows to change the schema of VMAPI's moray buckets where it's defined and have VMAPI take care of properly updating the buckets that changed without having to write ad-hoc scripts such as `tools/migrations/add-docker-index.js`. As a result, this PR removes the `tools/migrations` directory, since it only contained the `tools/migrations/add-docker-index.js` file.

It does not add any additional support for migrations though, that would need to be added in a future change.

The current state of this PR prevents VMAPI from starting up if an update for a given bucket fails. The update would happen when the service starts, not when the VMAPI zone is updated to a new image via sdcadm. I'm not sure this is acceptable.

It also always reindexes all VMAPI buckets so that all indexed fields are properly populated in the event of an error during a previous update.

### CODE REVIEW

File: lib/apis/moray.js [hunk 1]
```diff
@@ -145,12 +138,15 @@ Moray.prototype.connect = function () {
             log.error(err, 'moray client error');
         });
 
+        self.log.info('Setting up buckets...');
+
         self._setupBuckets(function (err) {
             if (err) {
-                self.log.error({ err: err }, 'Buckets were not loaded');
+                self.emit('error', err);
+                self.log.error('Error when setting up moray buckets: ' + err);
```
Comment (jclulow) [inline lib/apis/moray.js]: You should pass the error object as the first argument, e.g. `self.log.error(err, 'failed to set up moray buckets')`.  I would also do this _before_ emitting the error through the `EventEmitter`.

**Comment by trentm** [comment]
> As a result, this PR removes the tools/migrations directory, since it only contained the tools/migrations/add-docker-index.js file.

Is that assuming that all current VMAPI deployments have already upgraded to a VMAPI and added that index? (Naive question... I haven't read the commits yet.)

**Comment by misterdjules** [comment]
@trentm 

> > As a result, this PR removes the tools/migrations directory, since it only contained the tools/migrations/add-docker-index.js file.

> Is that assuming that all current VMAPI deployments have already upgraded to a VMAPI and added that index? (Naive question... I haven't read the commits yet.)

This change would make a VMAPI deployment that has not already added that index add it because the version of the `vmapi_vms` bucket would be 0, and this change:

1. sets the version to 1
2. adds a `'docker'` index

One of the things that is missing from this PR is a set of tests e.g for the use case you mentioned. I'm just still not sure how to write them without writing a lot of code, and so I thought I would gather some feedback about the intent of the change before that.

File: lib/apis/moray.js (hunk 1 above)
Comment (misterdjules) [inline lib/apis/moray.js]: Agreed for both points. Thanks!

File: lib/apis/moray.js [hunk 2]
```diff
@@ -1021,4 +1149,36 @@ Moray.prototype.delVmRoleTags = function (uuid, cb) {
...
+Moray.prototype._reindexBucket = function _reindexBucket(bucketName, callback) {
+    assert.string(bucketName, 'bucketName');
+    assert.func(callback, 'callback');
+
+    var self = this;
+
+    var rowsPerCall = 100;
+    var processed = rowsPerCall;
+
+    async.whilst(
```
Comment (jclulow) [inline lib/apis/moray.js]: I would ditch the usage of `async.whilst()` here, and just call yourself repeatedly.  e.g.

```javascript
Moray.prototype._reindexBucket = function _reindexBucket(bucketName, callback)
{
    assert.string(bucketName, 'bucketName');
    assert.func(callback, 'callback');

    var self = this;

    self.connection.reindexObjects(bucketName, 100, { noCache: true },
      function (err, res) {
        if (err || res.processed < 1) {
            callback(err);
            return;
        }

        self._reindexBucket(bucketName, callback);
    });
};
```

**Comment by jclulow** [comment]
I have a general SDC-level question.  If the `sdcadm update vmapi` process succeeds, but the VMAPI boot fails (because of a database schema versioning problem):

- How is the operator expected to know that this has happened?
- What is the operator supposed to do next?

**Comment by misterdjules** [comment]
@jclulow 

> I have a general SDC-level question. If the sdcadm update vmapi process succeeds, but the VMAPI boot fails (because of a database schema versioning problem):

> How is the operator expected to know that this has happened?
What is the operator supposed to do next?

That's exactly what I meant by:

> The current state of this PR prevents VMAPI from starting up if an update for a given bucket fails. The update would happen when the service starts, not when the VMAPI zone is updated to a new image via sdcadm. I'm not sure this is acceptable.

I'm not familiar with the mechanisms we have in place to deal with/be notified of these errors when the service starts. If we have no good story around that then I would think that either:

* we want to handle moray buckets' configuration changes at sdcadm update time.
* we need to add notification/diagnostic/repair tools for this use case.

That's where I need some guidance. FWIW, sdc-docker [uses a similar mechanism to upgrade moray buckets](https://github.com/joyent/sdc-docker/blob/master/lib/moray.js#L139-L213). Of course it doesn't mean this is the right way to do it.

File: lib/apis/moray.js (hunk 2 above)
Comment (misterdjules) [inline lib/apis/moray.js]: Sounds good, will do, thanks!

**Comment by trentm** [comment]
`sdcadm up vmapi` (and similarly for other insts) will effectively do a health check on that instance before saying "yes the upgrade was successful". The next Q is how good that health check is. Currently we still use `sdc-healthcheck` instead of `sdcadm check-health`, so that tells  you something.

The current healthcheck for a vmapi inst is whether its `/ping` endpoint succeeds. Which might mean any of:

1. always get a success while VMAPI is still trying to upgrade the buckets (dunno how VMAPI handles /ping when migrating); or
2. sometimes get a fialure for a slow migration (e.g. a race)

Would take testing with slow migrations, etc. And possibly long wanted improvements to sdcadm's health checking.

**Comment by jclulow** [comment]
On 5 January 2016 at 17:01, Trent Mick <notifications@github.com> wrote:
> The current healthcheck for a vmapi inst is whether its /ping endpoint succeeds. Which might mean any of:
>
> always get a success while VMAPI is still trying to upgrade the buckets (dunno how VMAPI handles /ping when migrating); or
> sometimes get a fialure for a slow migration (e.g. a race)

We should probably:

  (1) not respond to Ping requests until we're actually available

  (2) not fail an "sdcadm update" just because something takes too long


I think (1) is definitely in scope for this change, but (2) is
obviously going to be more work and in a different repository.



Cheers.

-- 
Joshua M. Clulow
Software Engineer @ Joyent
mail: jmc@joyent.com

**Comment by jclulow** [comment]
So, another potential issue.  Does this mean that once you deploy a VMAPI that includes a schema modification, you can never roll back to a previous version?

**Comment by misterdjules** [comment]
@jclulow 

> So, another potential issue. Does this mean that once you deploy a VMAPI that includes a schema modification, you can never roll back to a previous version?

Not without manual intervention (dropping the `_rver` column in the corresponding moray bucket(s) and resetting the version property in the bucket's config).

**Comment by jclulow** [comment]
> Not without manual intervention (dropping the `_rver` column in the corresponding moray bucket(s) and resetting the version property in the bucket's config).

So I think that's actually not what we want, then.  I've been discussing (with @melloc, amongst others) the idea of a _deferred update_ to SDC.  This is somewhat similar to feature flags, but they're more like fuses in that they only go one way.  In summary, you would have:

- a property in SAPI, which presumably translates into configuration via {{config-agent}}; this records whether or not the new feature has been enabled yet
- some set of preconditions for enabling the feature (e.g. "all instances of VMAPI are capable of using the feature", or "all vm-agent instances have been upgraded to a sufficiently recent version"), expressed programatically in SAPI
- some administrative interface (`sdcadm`, AdminUI) for expressing the deferred update to the operator, and what it means to turn it on (e.g. "you will no longer be able to downgrade VMAPI once you apply this schema update")
- no way to turn the thing back off once you've upgraded, but also no _requirement_ that you enable it for the time being, so that you can coordinate crossing what is essentially a kind of flag day
- new deployments will start out with all of these updates already "applied"

The goal is to turn implicit, irreversible schema upgrades into a coordinated event that the operator can control.  This functionality would ideally not be reinvented each time we need it, and would provide a uniform experience to the operator.

**Comment by misterdjules** [comment]
I just realized that when extracting the moray buckets' schema changes from https://github.com/joyent/sdc-vmapi/pull/16, I kept the code that bumps the buckets' version number, but it's actually not needed for what I need right now.

What I actually need is just to be able to add (and maybe remove) indexes without having to write an ad-hoc script such as `tools/migrations/add-docker-index.js` that needs to be run manually. I pushed an updated version of this PR that gets rid of the versioning changes while keeping the rest at https://github.com/misterdjules/sdc-vmapi/commit/7317a605d08b582ab4a548e7887b40b96326ddff.

Because these changes explicitly ignore the addition of version numbers to the schema, they seem safe to me: a rollback after adding/removing one or several index is always possible because:

1. there's no versioning on the bucket.
2. removing/adding an index does not remove data.

Now because the buckets' version number is not updated when adding/removing an index, [the fix to refresh the bucket schema cache on write when the schema changed](https://github.com/joyent/moray/commit/8f0112a) would not work. 

It means that the service would need to _not_ be available (not respond to /ping requests) while indexes are updated, otherwise we'd risk having newly created entries to not update the index. I would think that it would be ok to make the service available while reindexing is in progress, since new entries in moray would write to the index properly, and existing entries could be backfilled asynchronously.

@jclulow @trentm I'd like to get your feedback on whether this sounds like a reasonable path forward for the short term, without adding technical debt for long term goals.

**Comment by misterdjules** [comment]
@trentm @jclulow I had actually assumed that `tools/migrations/add-docker-index.js` was run manually, but [it's part of sdcadm update-other](https://github.com/joyent/sdcadm/blob/master/lib/cli/do_update_other.js#L759-L770). So I imagine I could actually add any index in the same way. Thoughts?

**Comment by misterdjules** [comment]
Closing, as this clearly needs to be better thought out. I'll try to come up with a proposal (as an RFD) that includes feedback from @trentm and @jclulow reasonably soon.

**Pull request closed without merging**